### DISCUSSION TITLE
TextFSM template help

### DISCUSSION THREAD

**Discussion by amitsova** [networktocode/ntc-templates#1578] [Q&A] [answered]
Hi all,

I am struggling to create a template which works for Cisco's "show interfaces trunk" command. Can someone share some hints? I tried a lot of regex variations in a combination with the textfsm options with no success. This is the output:

```
Port        Mode             Encapsulation  Status        Native vlan
Fa1/8       on               802.1q         trunking      1
Gi1/1       on               802.1q         trunking      1
Gi1/2       on               802.1q         trunking      1

Port        Vlans allowed on trunk
Fa1/8       1-4094
Gi1/1       1-4094
Gi1/2       1-4094

Port        Vlans allowed and active in management domain
Fa1/8       1,105,110,130,1010
Gi1/1       1,105,110,130,1010
Gi1/2       1,105,110,130,1010

Port        Vlans in spanning tree forwarding state and not pruned
Fa1/8       1,105,110,130,1010
Gi1/1       1,105,110,130,1010
Gi1/2       1,105,110,130,1010
```
Thank you in advance!

**Comment by yevgeniy-voloshin** [comment]
Hi Adriana,
maybe this link will shed some light on your problem:
https://stackoverflow.com/questions/50396833/use-textfsm-to-find-allowed-vlans-on-trunk

**Reply by amitsova** [reply]
Hi Yevgeniy!

Thanks a lot! I have something similar:

```
Value PORT (\S+)
Value MODE (\S+)
Value ENCAPS (\S+)
Value STATUS (\S+)
Value NATIVE (\d+)
Value VLAN_ON_TRUNK (\S+)
Value VLAN_MGMT_DOMAIN (\S+)

Start
  ^Port\s+Mode\s+Encapsulation\s+Status\s+Native\s+vlan\s*$$ -> Entry1


  
Entry1
  ^${PORT}\s+${MODE}\s+${ENCAPS}\s+${STATUS}\s+${NATIVE}\s*$$ -> Record
  ^Port\s+Vlans\s+allowed\s+on\s+trunk -> Entry2
  
Entry2
  ^\S+\s+${VLAN_ON_TRUNK}\s*$$ -> Record
  ^Port\s+Vlans\s+allowed\s+and\s+active\s+in\s+management\s+domain\s*$$ -> Entry3
  
Entry3
  ^\S+\s+${VLAN_MGMT_DOMAIN}\s*$$ -> Record
  ^Port\s+Vlans\s+in\s+spanning\s+tree\s+forwarding\s+state\s+and\s+not\s+pruned -> End
```

My problem is that the output looks like this:

![image](https://github.com/networktocode/ntc-templates/assets/39456433/93b18b0f-c7ee-4fc9-a3ce-51ab790b909e)

And I don't know if there is the possibility to have all in one row. Maybe there's a way, but unfortunately, I don't know it.

**Comment by amitsova** [comment]
Hi Jose,

I was not able to see your comment in Github. I am quite busy atm, but I will give a try your suggestion. It looks interesting 🙂

Thanks a lot!

Kind regards,
Adriana

Adriana Mitsova

https://www.linkedin.com/in/adriana-mitsova-188606116/
________________________________

**Reply by jorlandobr** [reply, marked as answer]
Hi Adriana!

I missed your reply and, apparently, deleted my initial reply by accident.

I remember (vaguely) about having some more success grouping the collected data using lists.

It seems that these modifications can give the intend data (grouped as lists). The extra label idea {Context} I got from this link, since when testing I could group all the sections, except the last one:

https://stackoverflow.com/questions/54649429/is-there-a-way-to-record-multiple-lines-at-once-with-textfsm

So I assumed that one extra step would do the trick.

```
Value Context (\S+\s\d+)
Value List PORT (\S+)
Value List MODE (\S+)
Value List ENCAPS (\S+)
Value List STATUS (\S+)
Value List NATIVE (\d+)
Value List VLAN_ALLOW (\S+)
Value List VLAN_MGMT_DOMAIN (\S+)
Value List VLAN_SPTF_PRUN (\S+)

Start
  ^Port\s+Mode\s+Encapsulation\s+Status\s+Native\s+vlan\s*$$
  ^${PORT}\s+${MODE}\s+${ENCAPS}\s+${STATUS}\s+${NATIVE}\s*$$ -> Continue
  ^Port\s+Vlans\s+allowed\s+on\s+trunk -> Entry1
  
Entry1
  ^\S+\s+${VLAN_ALLOW}\s*$$ -> Continue
  ^Port\s+Vlans\s+allowed\s+and\s+active\s+in\s+management\s+domain\s*$$ -> Entry2
  
Entry2
  ^\S+\s+${VLAN_MGMT_DOMAIN}\s*$$ -> Continue
  ^Port\s+Vlans\s+in\s+spanning\s+tree\s+forwarding\s+state\s+and\s+not\s+pruned -> Entry3

Entry3
  ^\S+\s+${VLAN_SPTF_PRUN}\s*$$ -> Continue
  ^${Context}: -> Start

```
To my tests, I modified a little your original input, to be sure that everithing was being captured and in order. The modified test data is this one:

```
Port        Mode             Encapsulation  Status        Native vlan
Fa1/8       on18               802.18q         trunking18      18
Gi1/1       on11               802.11q         trunking11      11
Gi1/2       on12               802.12q         trunking12      12

Port        Vlans allowed on trunk
Fa1/8       18-4094
Gi1/1       11-4094
Gi1/2       12-4094

Port        Vlans allowed and active in management domain
Fa1/8       18,105,110,130,1010
Gi1/1       11,105,110,130,1010
Gi1/2       12,105,110,130,1010

Port        Vlans in spanning tree forwarding state and not pruned
Fa1/8       1,105,110,138,1018
Gi1/1       1,105,110,131,1011
Gi1/2       1,105,110,132,1012
```

And after proccessed gives that answer that, I think, is what you need:

```
{
	"Context": "",
	"ENCAPS": [
		"802.18q",
		"802.11q",
		"802.12q"
	],
	"MODE": [
		"on18",
		"on11",
		"on12"
	],
	"NATIVE": [
		"18",
		"11",
		"12"
	],
	"PORT": [
		"Fa1/8",
		"Gi1/1",
		"Gi1/2"
	],
	"STATUS": [
		"trunking18",
		"trunking11",
		"trunking12"
	],
	"VLAN_ALLOW": [
		"18-4094",
		"11-4094",
		"12-4094"
	],
	"VLAN_MGMT_DOMAIN": [
		"18,105,110,130,1010",
		"11,105,110,130,1010",
		"12,105,110,130,1010"
	],
	"VLAN_SPTF_PRUN": [
		"1,105,110,138,1018",
		"1,105,110,131,1011",
		"1,105,110,132,1012"
	]
}
```

Hope this helps!

**Reply by jorlandobr** [reply]
By the way, each time you do a Record action the values read at that point are recorded and emptied and a new state is created, so that was the cause of the duplicate and empty fields as your script went down.

**Reply by mjbear** [reply]
@amitsova
Did Jose's suggestion solve for your goal?

**Reply by amitsova** [reply]
Hi @mjbear ,

So far, yes! :)

**Reply by mjbear** [reply]
@amitsova
Great!
Thank you for replying.
